src/tools/cephfs/JournalTool.cc

```diff
@@ -492,6 +539,243 @@ int JournalTool::journal_reset()
+      write_fnode = true;
+      // Note: creating the dirfrag *without* a backtrace, relying on
+      // MDS to regenerate backtraces in FSCK
+    } else if (r == 0) {
+      // Conditionally update existing omap header
+      fnode_t old_fnode;
+      bufferlist::iterator old_fnode_iter = old_fnode_bl.begin();
+      old_fnode.decode(old_fnode_iter);  // FIXME handle decode error and overwrite
+      dout(4) << "frag " << frag_oid.name << " fnode old v" <<
+        old_fnode.version << " vs new v" << lump.fnode.version << dendl;
+      write_fnode = old_fnode.version < lump.fnode.version;
+    } else {
+      // Unexpected error
+      dout(4) << "failed to read OMAP header from directory fragment "
+        << frag_oid.name << " " << cpp_strerror(r) << dendl;
+      return r;
+    }
+
+    if (write_fnode && !dry_run) {
+      dout(4) << "writing fnode to omap header" << dendl;
+      bufferlist fnode_bl;
+      lump.fnode.encode(fnode_bl);
+      r = io.omap_set_header(frag_oid.name, fnode_bl);
+      if (r != 0) {
+        derr << "Failed to write fnode for frag object " << frag_oid.name << dendl;
+        return r;
+      }
+    }
+
+    // Try to get the existing dentry
+    list<ceph::shared_ptr<EMetaBlob::fullbit> > const &fb_list = lump.get_dfull();
+    for (list<ceph::shared_ptr<EMetaBlob::fullbit> >::const_iterator fbi = fb_list.begin(); fbi != fb_list.end(); ++fbi) {
+      EMetaBlob::fullbit const &fb = *(*fbi);
+
+      // Get a key like "foobar_head"
+      std::string key;
+      dentry_key_t dn_key(fb.dnlast, fb.dn.c_str());
+      dn_key.encode(key);
+
+      // See if the dentry is present
+      std::set<std::string> keys;
+      keys.insert(key);
+      std::map<std::string, bufferlist> vals;
+      r = io.omap_get_vals_by_keys(frag_oid.name, keys, &vals);
+      if (r != 0) {
+        derr << "unexpected error reading fragment object "
+             << frag_oid.name << ": " << cpp_strerror(r) << dendl;
+        return r;
+      }
+    
+      dout(4) << "inspecting fullbit " << frag_oid.name << "/" << fb.dn << dendl;
+      bool write_dentry = false;
+      if (vals.find(key) == vals.end()) {
+        dout(4) << "dentry did not already exist, will create" << dendl;
+        write_dentry = true;
+      } else {
+        dout(4) << "dentry " << key << " existed already" << dendl;
+        assert(vals.size() == 1);
+        dout(4) << "dentry exists, checking versions..." << dendl;
+        bufferlist &old_dentry = vals[key];
+        // Decode dentry+inode
+        bufferlist::iterator q = old_dentry.begin();
+
+        snapid_t dnfirst;
+        ::decode(dnfirst, q);
+        char dentry_type;
+        ::decode(dentry_type, q);
+
+        if (dentry_type == 'L') {
+          // leave write_dentry false, we have no version to
+          // compare with in a hardlink, so it's not safe to
+          // squash over it with what's in this fullbit
+          dout(4) << "skipping hardlink dentry in backing store" << dendl;
```

Comment:
Urgh, this needs to do something a little more useful, at lest telling us what dentry is the problem.
And if we don't have a version associated with hardlinks (really? can't we just go by the directory version or something?) then we should probably consider adding one.